File: scripts/tictactoe/tictactoe.py
import random, time

class TicTacToe:
    def __init__(self):
        self.board = [['-' for _ in range(3)] for _ in range(3)]
        self.current_player = 'X'

    def switch_player(self):
        self.current_player = 'O' if self.current_player == 'X' else 'X'

    def check_win(self, player) -> bool:
        for i in range(3):
            if all(self.board[i][j] == player for j in range(3)):
                return True
            if all(self.board[j][i] == player for j in range(3)):
                return True
        if all(self.board[i][i] == player for i in range(3)):
            return True
        if all(self.board[i][2 - i] == player for i in range(3)):
            return True
        return False

    def is_full(self):
        return all(self.board[i][j] != '-' for i in range(3) for j in range(3))

    def get_random_move(self):
        available_moves = [(i, j) for i in range(3) for j in range(3) if self.board[i][j] == '-']
        return random.choice(available_moves)

    def display_board(self):
        for row in self.board:
            print(' '.join(row))
        print()

    def play_game(self):
        while True:
            move = self.get_random_move()
            self.board[move[0]][move[1]] = self.current_player
            self.display_board()
            if self.check_win(self.current_player):
                print(f"Player {self.current_player} wins!\n")
                return self.current_player
            if self.is_full():
                print("It's a draw!\n")
                return '-'
            self.switch_player()

if __name__ == "__main__":
    shit: int = 100
    x_wins, o_wins, draws = 0, 0, 0

    start_time = time.time()

    for i in range(shit):
        print(f"Game {i + 1}")
        game = TicTacToe()
        result = game.play_game()
        if result == 'X':
            x_wins += 1
        elif result == 'O':
            o_wins += 1
        else:
            draws += 1

    end_time = time.time()
    total_time = end_time - start_time

    print(f"""Final Results:
    X Wins: {x_wins}
    O Wins: {o_wins}
    Draws: {draws}
    Total Time: {1000 * total_time:.3f} seconds
    """)

File: scripts/tictactoe/tictactoe.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <chrono>

using namespace std;
using namespace chrono;

class TicTacToe {
private:
    vector<vector<char>> board;
    char currentPlayer;

    void switchPlayer() {
        currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
    }

    bool checkWin(char player) {
        for (int i = 0; i < 3; i++) {
            if (board[i][0] == player && board[i][1] == player && board[i][2] == player) return true;
            if (board[0][i] == player && board[1][i] == player && board[2][i] == player) return true;
        }
        if (board[0][0] == player && board[1][1] == player && board[2][2] == player) return true;
        if (board[0][2] == player && board[1][1] == player && board[2][0] == player) return true;
        return false;
    }

    bool isFull() {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[i][j] == '-') return false;
            }
        }
        return true;
    }

    pair<int, int> getRandomMove() {
        vector<pair<int, int>> availableMoves;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[i][j] == '-') availableMoves.emplace_back(i, j);
            }
        }
        return availableMoves[rand() % availableMoves.size()];
    }

    void displayBoard() {
        for (const auto& row : board) {
            for (char cell : row) {
                cout << cell << " ";
            }
            cout << "\n";
        }
        cout << "\n";
    }

public:
    TicTacToe() {
        board = vector<vector<char>>(3, vector<char>(3, '-'));
        currentPlayer = 'X';
    }

    char playGame() {
        while (true) {
            pair<int, int> move = getRandomMove();
            board[move.first][move.second] = currentPlayer;
            displayBoard();
            if (checkWin(currentPlayer)) {
                cout << "Player " << currentPlayer << " wins L bozo!\n\n";
                return currentPlayer;
            }
            if (isFull()) {
                cout << "Damn a draw! Skill issue lol\n\n";
                return '-';
            }
            switchPlayer();
        }
    }
};

int main() {
    srand(time(0));
    int games = 100;
    int xWins = 0, oWins = 0, draws = 0;

    auto startTime = high_resolution_clock::now();

    for (int i = 0; i < games; i++) {
        cout << "Game " << i + 1 << "\n";
        TicTacToe game;
        char result = game.playGame();
        if (result == 'X') xWins++;
        else if (result == 'O') oWins++;
        else draws++;
    }

    auto endTime = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(endTime - startTime);

    cout << "Final Results:\n";
    cout << "X Wins: " << xWins << "\n";
    cout << "O Wins: " << oWins << "\n";
    cout << "Draws: " << draws << "\n";
    cout << "Total Time: " << duration.count() << " ms\n";

    return 0;
}